packages/story-editor/src/app/taxonomy/taxonomyProvider.js

```diff
@@ -0,0 +1,205 @@
+import Context from './context';
+import {
+  dictonaryOnKey,
+  mapObjectVals,
+  mergeNestedDictionaries,
+  objectFromEntries,
+  mapObjectKeys,
+  cacheFromEmbeddedTerms,
+} from './utils';
+
+function TaxonomyProvider(props) {
+  const [taxonomies, setTaxonomies] = useState([]);
+  const [selectedSlugs, setSelectedSlugs] = useState({});
+  const [termCache, setTermCache] = useState({});
+  const { updateStory, isStoryLoaded, story } = useStory(
+    ({ state: { pages, story }, actions: { updateStory } }) => ({
+      updateStory,
+      isStoryLoaded: pages.length > 0,
+      story,
+    })
+  );
+
+  const { getTaxonomyTerm, createTaxonomyTerm, getTaxonomies } = useAPI(
+    ({ actions }) => actions
+  );
+
+  // Get all registered `web-story` taxonomies.
+  useEffect(() => {
+    (async function () {
+      try {
+        const result = await getTaxonomies();
+        setTaxonomies(Object.values(result));
+      } catch (e) {
+        // Do we wanna do anything here?
+      }
+    })();
+  }, [getTaxonomies]);
+
+  // Reference embedded terms in the story and taxonomies
+  // to get the initial selected terms as well as populate
+  // the taxonomy term cache
+  const hasHydratedRunOnce = useRef(false);
+  useEffect(() => {
+    if (taxonomies.length > 0 && isStoryLoaded && !hasHydratedRunOnce.current) {
+      const taxonomiesBySlug = dictonaryOnKey(taxonomies, 'slug');
+      const initialCache = mapObjectKeys(
+        cacheFromEmbeddedTerms(story.embeddedTerms),
+        (slug) => taxonomiesBySlug[slug].rest_base
+      );
+      const initialSelectedSlugs = mapObjectVals(initialCache, (val) =>
+        Object.keys(val)
+      );
+
+      hasHydratedRunOnce.current = true;
+      setTermCache(initialCache);
+      setSelectedSlugs(initialSelectedSlugs);
+    }
+  }, [story, isStoryLoaded, taxonomies, setSelectedSlugs, setTermCache]);
+
+  // With the freeform taxonomy input, we can have terms selected
+  // that may be in the process of being created or retrieved from
+  // the backend. Becuase of this, we sync up our local selected slugs
+  // with whatever cached terms are available at any given moment.
+  useEffect(() => {
+    if (!hasHydratedRunOnce.current) {
+      return;
+    }
+
+    const termEntries = Object.entries(selectedSlugs).map(
+      ([taxonomyRestBase, termSlugs = []]) => [
+        taxonomyRestBase,
+        termSlugs
+          .map((termSlug) => termCache[taxonomyRestBase]?.[termSlug]?.id)
+          .filter((id) => typeof id === 'number'),
+      ]
+    );
+    const terms = objectFromEntries(termEntries);
+    updateStory({
+      properties: terms,
+    });
+  }, [updateStory, selectedSlugs, termCache]);
+
+  const addSearchResultsToCache = useCallback(
+    async (taxonomy, name) => {
+      let response = [];
+      try {
+        response = await getTaxonomyTerm(taxonomy.rest_base, {
+          search: name,
+          // This is the per_page value Gutenberg is using
+          per_page: 20,
```

Comment:
💭 maybe we can set this as a default in the options but allow a request to be sent with more (100 is the max right?)

Reply:
Yup, 100 is the max, and WP sets the default to 10 if you don't specify.

We can update the function to make the default 20, but feels kinda like arbitrarily moving things around since I'm thinking the value of 20 results per page will be unique behavior to the scope of this function.

Also kinda wanted to minimize making `getTaxonomyTerm` opinionated over the actual query until we see any patterns arise from our use of it.

anyway, happy to update, don't feel strongly either way, but just wanted to shed some light on my rationale.

Reply:
makes sense - let's keep it!